### ISSUE TITLE
HTTP/2 with prior knowledge?

### ISSUE THREAD

**Issue by greenday9** [warmcat/libwebsockets#1890]
I'm using the latest master commit of libwebsockets on a Centos 7.x platform.

Is there a way to support HTTP/2 with prior knowledge over clear text connections (client and server)?  In that case, we would just start with the connection preface over a connection with no upgrade request.   From what I've tested, it seems like upgrade request is required for libwebsockets to get to HTTP/2 on cleartext.

**Comment by lws-team** [comment]
Right it doesn't support it atm.

It supports discovery by alpn which is by far the most common way, and upgrade.

If you want to add it, a patch is welcome... the wsi operate in a mixed generic / protocol-specific state machine

https://libwebsockets.org/git/libwebsockets/tree/lib/roles/private-lib-roles.h#n85

The generic parts make it go through the right sequence of socks5 if that's enabled, tls negotiation, protocol-specific handshake like h2 preface, connection upgrade (eg, ws) and finally to be established.

You'd want a vhost-specific option to tell the vhost he's serving h2-with-prior-knowledge mode, and shortcut accepted connections to wsi->state LRS_H2_AWAIT_PREFACE like it does after tls / alpn the other way

https://libwebsockets.org/git/libwebsockets/tree/lib/roles/h2/ops-h2.c#n1165

For h2 client you'd do something similar, he already knows to send the preface later, you'd shortcut him to do it earlier

https://libwebsockets.org/git/libwebsockets/tree/lib/roles/h2/http2.c#n363

Basically it already does all the right things you just need to add bits to reorder when they happen.

**Comment by greenday9** [comment]
Thanks for the quick response.  I'll take a look.

**Comment by greenday9** [comment]
Update.  I've gotten it to work with minor state changes.  Once I clean it up, I'll generate a pull request.

**Comment by lws-team** [comment]
Good news!

**Comment by lws-team** [comment]
Thanks a lot for the patch, it's on master now.
